File: source/jugglinglab/core/AnimationPrefs.java
// AnimationPrefs.java
//
// Copyright 2002-2022 Jack Boyce and the Juggling Lab contributors

package jugglinglab.core;

import java.awt.*;
import java.text.MessageFormat;
import java.util.ResourceBundle;
import java.util.StringTokenizer;

import jugglinglab.jml.JMLPattern;
import jugglinglab.util.*;
import jugglinglab.view.View;


public class AnimationPrefs {
    static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
    static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

    public static final int GROUND_AUTO = 0;  // must be sequential
    public static final int GROUND_ON = 1;  // starting from 0
    public static final int GROUND_OFF = 2;

    // default values of all items
    public static final int width_def = 400;
    public static final int height_def = 450;
    public static final double fps_def;
    public static final double slowdown_def = 2.0;
    public static final int border_def = 0;
    public static final int showGround_def = GROUND_AUTO;
    public static final boolean stereo_def = false;
    public static final boolean startPause_def = false;
    public static final boolean mousePause_def = false;
    public static final boolean catchSound_def = false;
    public static final boolean bounceSound_def;
    public static final int view_def = View.VIEW_NONE;

    static {
        // audio clip playback seems to block on Linux
        if (jugglinglab.JugglingLab.isLinux)
            bounceSound_def = false;
        else
            bounceSound_def = false;

        // set default `fps` to screen refresh rate, if possible
        double fps_screen = 0.0;

        try {
            GraphicsEnvironment env = GraphicsEnvironment.getLocalGraphicsEnvironment();
            GraphicsDevice[] devices = env.getScreenDevices();
            if (devices.length > 0)
                fps_screen = (double)devices[0].getDisplayMode().getRefreshRate();
                // getRefreshRate() returns 0 when refresh is unknown
        } catch (Exception e) {
            // HeadlessException when running headless (from CLI)
        }

        fps_def = (fps_screen < 20.0 ? 60.0 : fps_screen);
    }

    public int width = width_def;
    public int height = height_def;
    public double fps = fps_def;
    public double slowdown = slowdown_def;
    public int border = border_def;
    public int showGround = showGround_def;
    public boolean stereo = stereo_def;
    public boolean startPause = startPause_def;
    public boolean mousePause = mousePause_def;
    public boolean catchSound = catchSound_def;
    public boolean bounceSound = bounceSound_def;
    public double[] camangle;  // in degrees! null means use default
    public int view = view_def;  // one of the values in View
    public int[] hideJugglers;


    public AnimationPrefs() {
        super();
    }

    public AnimationPrefs(AnimationPrefs jc) {
        if (jc.width > 0)
            width = jc.width;
        if (jc.height > 0)
            height = jc.height;
        if (jc.slowdown >= 0.0)
            slowdown = jc.slowdown;
        if (jc.fps >= 0.0)
            fps = jc.fps;
        if (jc.border >= 0)
            border = jc.border;
        showGround = jc.showGround;
        stereo = jc.stereo;
        startPause = jc.startPause;
        mousePause = jc.mousePause;
        catchSound = jc.catchSound;
        bounceSound = jc.bounceSound;
        if (jc.camangle != null)
            camangle = jc.camangle.clone();
        view = jc.view;
        if (jc.hideJugglers != null)
            hideJugglers = jc.hideJugglers.clone();
    }

    public AnimationPrefs fromParameters(ParameterList pl) throws JuggleExceptionUser {
        int tempint;
        double tempdouble;
        String value = null;

        if ((value = pl.removeParameter("stereo")) != null)
            stereo = Boolean.parseBoolean(value);
        if ((value = pl.removeParameter("startpaused")) != null)
            startPause = Boolean.parseBoolean(value);
        if ((value = pl.removeParameter("mousepause")) != null)
            mousePause = Boolean.parseBoolean(value);
        if ((value = pl.removeParameter("catchsound")) != null)
            catchSound = Boolean.parseBoolean(value);
        if ((value = pl.removeParameter("bouncesound")) != null)
            bounceSound = Boolean.parseBoolean(value);
        if ((value = pl.removeParameter("fps")) != null) {
            try {
                tempdouble = Double.parseDouble(value);
                fps = tempdouble;
            } catch (NumberFormatException e) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "fps" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("slowdown")) != null) {
            try {
                tempdouble = Double.parseDouble(value);
                slowdown = tempdouble;
            } catch (NumberFormatException e) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "slowdown" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("border")) != null) {
            try {
                tempint = Integer.parseInt(value);
                border = tempint;
            } catch (NumberFormatException e) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "border" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("width")) != null) {
            try {
                tempint = Integer.parseInt(value);
                width = tempint;
            } catch (NumberFormatException e) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "width" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("height")) != null) {
            try {
                tempint = Integer.parseInt(value);
                height = tempint;
            } catch (NumberFormatException e) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "height" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("showground")) != null) {
            if (value.equalsIgnoreCase("auto"))
                showGround = GROUND_AUTO;
            else if (value.equalsIgnoreCase("true") || value.equalsIgnoreCase("on")
                        || value.equalsIgnoreCase("yes"))
                showGround = GROUND_ON;
            else if (value.equalsIgnoreCase("false") || value.equalsIgnoreCase("off")
                        || value.equalsIgnoreCase("no"))
                showGround = GROUND_OFF;
            else {
                String template = errorstrings.getString("Error_showground_value");
                Object[] arguments = { value };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("camangle")) != null) {
            try {
                double[] ca = new double[2];
                ca[1] = 90.0;        // default if second angle isn't given

                value = value.replace("(", "").replace(")", "");
                value = value.replace("{", "").replace("}", "");

                StringTokenizer st = new StringTokenizer(value, ",");
                int numangles = st.countTokens();
                if (numangles > 2) {
                    String template = errorstrings.getString("Error_too_many_elements");
                    Object[] arguments = { "camangle" };
                    throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
                }

                for (int i = 0; i < numangles; i++)
                    ca[i] = Double.parseDouble(st.nextToken().trim());

                camangle = new double[2];
                camangle[0] = ca[0];
                camangle[1] = ca[1];
            } catch (NumberFormatException e) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "camangle" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("view")) != null) {
            view = -1;
            for (int view_index = 0; view_index < View.viewNames.length; view_index++)
                if (value.equalsIgnoreCase(View.viewNames[view_index]))
                    view = view_index + 1;
            if (view == -1) {
                String template = errorstrings.getString("Error_unrecognized_view");
                Object[] arguments = { "'" + value + "'" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        if ((value = pl.removeParameter("hidejugglers")) != null) {
            value = value.replace("(", "").replace(")", "");

            StringTokenizer st = new StringTokenizer(value, ",");
            int numjugglers = st.countTokens();
            hideJugglers = new int[numjugglers];

            try {
                for (int i = 0; i < numjugglers; i++)
                    hideJugglers[i] = Integer.parseInt(st.nextToken().trim());
            } catch (NumberFormatException e) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "hidejugglers" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
        return this;
    }

    public AnimationPrefs fromString(String s) throws JuggleExceptionUser {
        ParameterList pl = new ParameterList(s);
        fromParameters(pl);
        pl.errorIfParametersLeft();
        return this;
    }

    public Dimension getSize() {
        return new Dimension(width, height);
    }

    public void setSize(Dimension dim) {
        width = dim.width;
        height = dim.height;
    }

    @Override
    public String toString() {
        String result = "";

        if (width != width_def)
            result += "width=" + width + ";";
        if (height != height_def)
            result += "height=" + height + ";";
        if (fps != fps_def)
            result += "fps=" + JLFunc.toStringRounded(fps, 2) + ";";
        if (slowdown != slowdown_def)
            result += "slowdown=" + JLFunc.toStringRounded(slowdown, 2) + ";";
        if (border != border_def)
            result += "border=" + border + ";";
        if (showGround != showGround_def) {
            switch (showGround) {
                case GROUND_AUTO:
                    result += "showground=auto;";
                    break;
                case GROUND_ON:
                    result += "showground=true;";
                    break;
                case GROUND_OFF:
                    result += "showground=false;";
                    break;
            }
        }
        if (stereo != stereo_def)
            result += "stereo=" + stereo + ";";
        if (startPause != startPause_def)
            result += "startpaused=" + startPause + ";";
        if (mousePause != mousePause_def)
            result += "mousepause=" + mousePause + ";";
        if (catchSound != catchSound_def)
            result += "catchsound=" + catchSound + ";";
        if (bounceSound != bounceSound_def)
            result += "bouncesound=" + bounceSound + ";";
        if (camangle != null)
            result += "camangle=(" + camangle[0] + "," + camangle[1] + ");";
        if (view != view_def)
            result += "view=" + View.viewNames[view - 1] + ";";
        if (hideJugglers != null) {
            result += "hidejugglers=(";
            for (int i = 0; i < hideJugglers.length; i++) {
                result += Integer.toString(hideJugglers[i]);
                if (i != hideJugglers.length - 1)
                    result += ",";
            }
            result += ");";
        }

        if (result.length() != 0)
            result = result.substring(0, result.length() - 1);

        return result;
    }
}

File: src/commonMain/kotlin/jugglinglab/core/AnimationPrefs.kt
//
// AnimationPrefs.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

@file:Suppress("SimplifyBooleanWithConstants")

package jugglinglab.core

import jugglinglab.generated.resources.*
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.ParameterList
import jugglinglab.util.getScreenFps
import jugglinglab.util.getStringResource
import jugglinglab.util.jlToStringRounded
import androidx.compose.ui.unit.IntSize

class AnimationPrefs {
    var width: Int = WIDTH_DEF
    var height: Int = HEIGHT_DEF
    var fps: Double = FPS_DEF
    var slowdown: Double = SLOWDOWN_DEF
    var border: Int = BORDER_DEF
    var showGround: Int = SHOWGROUND_DEF
    var stereo: Boolean = STEREO_DEF
    var startPause: Boolean = STARTPAUSE_DEF
    var mousePause: Boolean = MOUSEPAUSE_DEF
    var catchSound: Boolean = CATCHSOUND_DEF
    var bounceSound: Boolean = BOUNCESOUND_DEF
    var camangle: DoubleArray? = null // in degrees! null means use default
    var view: Int = VIEW_DEF // one of the values in View
    var hideJugglers: IntArray? = null

    constructor() : super()

    constructor(jc: AnimationPrefs) {
        if (jc.width > 0) {
            width = jc.width
        }
        if (jc.height > 0) {
            height = jc.height
        }
        if (jc.slowdown >= 0) {
            slowdown = jc.slowdown
        }
        if (jc.fps >= 0) {
            fps = jc.fps
        }
        if (jc.border >= 0) {
            border = jc.border
        }
        showGround = jc.showGround
        stereo = jc.stereo
        startPause = jc.startPause
        mousePause = jc.mousePause
        catchSound = jc.catchSound
        bounceSound = jc.bounceSound
        if (jc.camangle != null) {
            camangle = jc.camangle!!.clone()
        }
        view = jc.view
        if (jc.hideJugglers != null) {
            hideJugglers = jc.hideJugglers!!.clone()
        }
    }

    @Throws(JuggleExceptionUser::class)
    fun fromParameters(pl: ParameterList): AnimationPrefs {
        var tempint: Int
        var tempdouble: Double
        var value: String?

        if ((pl.removeParameter("stereo").also { value = it }) != null) {
            stereo = value.toBoolean()
        }
        if ((pl.removeParameter("startpaused").also { value = it }) != null) {
            startPause = value.toBoolean()
        }
        if ((pl.removeParameter("mousepause").also { value = it }) != null) {
            mousePause = value.toBoolean()
        }
        if ((pl.removeParameter("catchsound").also { value = it }) != null) {
            catchSound = value.toBoolean()
        }
        if ((pl.removeParameter("bouncesound").also { value = it }) != null) {
            bounceSound = value.toBoolean()
        }
        if ((pl.removeParameter("fps").also { value = it }) != null) {
            try {
                tempdouble = value!!.toDouble()
                fps = tempdouble
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "fps")
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("slowdown").also { value = it }) != null) {
            try {
                tempdouble = value!!.toDouble()
                slowdown = tempdouble
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "slowdown")
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("border").also { value = it }) != null) {
            try {
                tempint = value!!.toInt()
                border = tempint
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "border")
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("width").also { value = it }) != null) {
            try {
                tempint = value!!.toInt()
                width = tempint
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "width")
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("height").also { value = it }) != null) {
            try {
                tempint = value!!.toInt()
                height = tempint
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "height")
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("showground").also { value = it }) != null) {
            showGround = if (value.equals("auto", ignoreCase = true)) {
                GROUND_AUTO
            } else if (value.equals("true", ignoreCase = true)
                || value.equals("on", ignoreCase = true)
                || value.equals("yes", ignoreCase = true)
            ) {
                GROUND_ON
            } else if (value.equals("false", ignoreCase = true)
                || value.equals("off", ignoreCase = true)
                || value.equals("no", ignoreCase = true)
            ) {
                GROUND_OFF
            } else {
                val message = getStringResource(Res.string.error_showground_value, value)
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("camangle").also { value = it }) != null) {
            try {
                val ca = DoubleArray(2)
                ca[1] = 90.0  // default if second angle isn't given

                val tokens = value!!.replace(Regex("[(){}]"), "").split(',')
                if (tokens.size > 2) {
                    val message = getStringResource(Res.string.error_too_many_elements, "camangle")
                    throw JuggleExceptionUser(message)
                }
                tokens.forEachIndexed { i, token ->
                    if (token.isNotBlank()) {
                        ca[i] = token.trim().toDouble()
                    }
                }
                camangle = ca
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "camangle")
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("view").also { value = it }) != null) {
            view = -1
            for (viewIndex in viewNames.indices) {
                if (value.equals(viewNames[viewIndex], ignoreCase = true)) {
                    view = viewIndex + 1
                }
            }

            if (view == -1) {
                val message = getStringResource(Res.string.error_unrecognized_view, value)
                throw JuggleExceptionUser(message)
            }
        }
        if ((pl.removeParameter("hidejugglers").also { value = it }) != null) {
            try {
                val tokens = value!!.replace(Regex("[()]"), "").split(',')
                hideJugglers = tokens.mapNotNull { token ->
                    token.trim().takeIf { it.isNotEmpty() }?.toInt()
                }.toIntArray()
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "hidejugglers")
                throw JuggleExceptionUser(message)
            }
        }
        return this
    }

    @Throws(JuggleExceptionUser::class)
    fun fromString(s: String?): AnimationPrefs {
        val pl = ParameterList(s)
        fromParameters(pl)
        pl.errorIfParametersLeft()
        return this
    }

    var size: IntSize
        get() = IntSize(width, height)
        set(dim) {
            width = dim.width
            height = dim.height
        }

    @Suppress("KotlinConstantConditions")
    override fun toString(): String {
        val sb = StringBuilder()
        if (width != WIDTH_DEF) {
            sb.append("width=").append(width).append(";")
        }
        if (height != HEIGHT_DEF) {
            sb.append("height=").append(height).append(";")
        }
        if (fps != FPS_DEF) {
            sb.append("fps=").append(jlToStringRounded(fps, 2)).append(";")
        }
        if (slowdown != SLOWDOWN_DEF) {
            sb.append("slowdown=").append(jlToStringRounded(slowdown, 2)).append(";")
        }
        if (border != BORDER_DEF) {
            sb.append("border=").append(border).append(";")
        }
        if (showGround != SHOWGROUND_DEF) {
            when (showGround) {
                GROUND_AUTO -> sb.append("showground=auto;")
                GROUND_ON -> sb.append("showground=true;")
                GROUND_OFF -> sb.append("showground=false;")
            }
        }
        if (stereo != STEREO_DEF) {
            sb.append("stereo=").append(stereo).append(";")
        }
        if (startPause != STARTPAUSE_DEF) {
            sb.append("startpaused=").append(startPause).append(";")
        }
        if (mousePause != MOUSEPAUSE_DEF) {
            sb.append("mousepause=").append(mousePause).append(";")
        }
        if (catchSound != CATCHSOUND_DEF) {
            sb.append("catchsound=").append(catchSound).append(";")
        }
        if (bounceSound != BOUNCESOUND_DEF) {
            sb.append("bouncesound=").append(bounceSound).append(";")
        }
        if (camangle != null) {
            sb.append("camangle=(").append(camangle!![0]).append(",").append(camangle!![1]).append(");")
        }
        if (view != VIEW_DEF) {
            sb.append("view=").append(viewNames[view - 1]).append(";")
        }
        if (hideJugglers != null) {
            sb.append("hidejugglers=(")
            for (i in hideJugglers!!.indices) {
                sb.append(hideJugglers!![i])
                if (i != hideJugglers!!.size - 1) {
                    sb.append(",")
                }
            }
            sb.append(");")
        }
        if (!sb.isEmpty()) {
            sb.setLength(sb.length - 1)
        }
        return sb.toString()
    }

    companion object {
        // showground options
        const val GROUND_AUTO: Int = 0 // must be sequential
        const val GROUND_ON: Int = 1 // starting from 0
        const val GROUND_OFF: Int = 2

        // view options
        //
        // these should be sequential and in the same order as in the View menu,
        // because of assumptions in PatternWindow's constructor
        const val VIEW_NONE: Int = 0
        const val VIEW_SIMPLE: Int = 1
        const val VIEW_EDIT: Int = 2
        const val VIEW_PATTERN: Int = 3
        const val VIEW_SELECTION: Int = 4

        // in the same order as the VIEW_ constants above, used for `view`
        // parameter setting in AnimationPrefs
        val viewNames: List<String> = listOf(
            "simple",
            "visual_editor",
            "pattern_editor",
            "selection_editor",
        )

        // default values of all AnimationPrefs items
        const val WIDTH_DEF: Int = 400
        const val HEIGHT_DEF: Int = 450
        val FPS_DEF: Double = getScreenFps()
        const val SLOWDOWN_DEF: Double = 2.0
        const val BORDER_DEF: Int = 0
        const val SHOWGROUND_DEF: Int = GROUND_AUTO
        const val STEREO_DEF: Boolean = false
        const val STARTPAUSE_DEF: Boolean = false
        const val MOUSEPAUSE_DEF: Boolean = false
        const val CATCHSOUND_DEF: Boolean = false
        const val BOUNCESOUND_DEF: Boolean = false
        const val VIEW_DEF: Int = VIEW_NONE
    }
}
